cpp/src/arrow/json/parser_test.cc

```diff
@@ -144,14 +144,48 @@ TEST(BlockParserWithSchema, SkipFieldsOutsideSchema) {
                       "[\"thing\", null, \"\xe5\xbf\x8d\", null]"});
 }
 
-TEST(BlockParserWithSchema, FailOnInconvertible) {
-  auto options = ParseOptions::Defaults();
-  options.explicit_schema = schema({field("a", int32())});
-  options.unexpected_field_behavior = UnexpectedFieldBehavior::Ignore;
+class BlockParserTypeError : public ::testing::TestWithParam<UnexpectedFieldBehavior> {
+ public:
+  ParseOptions Options(std::shared_ptr<Schema> explicit_schema) {
+    auto options = ParseOptions::Defaults();
+    options.explicit_schema = std::move(explicit_schema);
+    options.unexpected_field_behavior = GetParam();
+    return options;
+  }
+};
+
+TEST_P(BlockParserTypeError, FailOnInconvertible) {
+  auto options = Options(schema({field("a", int32())}));
+  std::shared_ptr<Array> parsed;
+  Status error = ParseFromString(options, "{\"a\":0}\n{\"a\":true}", &parsed);
+  ASSERT_RAISES(Invalid, error);
+  ASSERT_EQ(error.message(),
+            "JSON parse error: Column(/a) changed from number to boolean in row 1");
+}
+
+TEST_P(BlockParserTypeError, FailOnNestedInconvertible) {
+  auto options = Options(schema({field("a", list(struct_({field("a", int32())})))}));
+  std::shared_ptr<Array> parsed;
+  Status error =
+      ParseFromString(options, "{\"a\":[{\"a\":0}]}\n{\"a\":[{\"a\":true}]}", &parsed);
+  ASSERT_RAISES(Invalid, error);
+  ASSERT_EQ(error.message(),
+            "JSON parse error: Column(/a/[]/a) changed from number to boolean in row 1");
```

Comment:
Test with two different field names perhaps? (so that we know that we get e.g. "/a/[]/b" and not "/b/[]/a")

Reply:
will do